ext/mbstring/mbstring.c

```diff
@@ -3068,55 +3065,77 @@ static const mbfl_encoding* mb_guess_encoding(unsigned char *in, size_t in_len,
+							length--;
+							if (length == 0) {
+								return 0;
+							}
+							memmove(&array[i], &array[i+1], (length - i) * sizeof(struct candidate));
 							goto try_next_encoding;
 						} else {
-							data[i].demerits += 1000;
+							array[i].demerits += 1000;
 						}
 					} else {
-						data[i].demerits += mb_estimate_encoding_demerits(w);
+						array[i].demerits += estimate_demerits(w);
 					}
 				}
-				if (data[i].in_len == 0) {
+				if (array[i].in_len == 0) {
 					finished++;
 				}
 			}
-			i++;
 		}
 	}
 
-	if (strict) {
+	return length;
+}
+
+MBSTRING_API const mbfl_encoding* mb_guess_encoding_for_strings(unsigned char **strings, size_t *str_lengths, size_t n, const mbfl_encoding **elist, unsigned int elist_size, bool strict)
+{
+	if (elist_size == 0) {
+		return NULL;
+	}
+	if (elist_size == 1) {
+		if (strict) {
+			while (n--) {
+				if (!php_mb_check_encoding((const char*)strings[n], str_lengths[n], *elist)) {
+					return NULL;
+				}
+			}
+		}
+		return *elist;
+	}
+	if (n == 1 && *str_lengths == 0) {
+		return *elist;
+	}
+
+	/* Allocate on stack; when we return, this array is automatically freed */
+	struct candidate *array = alloca(elist_size * sizeof(struct candidate));
+	elist_size = init_candidate_array(array, elist_size, elist, strings, str_lengths, n, strict);
+
+	while (n--) {
+		start_string(array, elist_size, strings[n], str_lengths[n]);
+		elist_size = count_demerits(array, elist_size, strict);
 		if (elist_size == 0) {
 			/* All candidates were eliminated */
 			return NULL;
 		}
-		/* The above loop might have broken because there was only 1 candidate encoding left
-		 * If in strict mode, we still need to process any remaining input for that candidate */
-		if (elist_size == 1 && data[0].in_len) {
-			const mbfl_encoding *enc = data[0].enc;
-			unsigned char *in = data[0].in;
-			size_t in_len = data[0].in_len;
-			unsigned int state = data[0].state;
-			while (in_len) {
-				size_t out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
-				while (out_len) {
-					if (wchar_buf[--out_len] == MBFL_BAD_INPUT) {
-						return NULL;
-					}
-				}
-			}
-		}
 	}
 
 	/* See which remaining candidate encoding has the least demerits */
 	unsigned int best = 0;
 	for (unsigned int i = 1; i < elist_size; i++) {
-		if (data[i].demerits < data[best].demerits) {
+		if (array[i].demerits < array[best].demerits) {
 			best = i;
 		}
 	}
-	return data[best].enc;
+	return array[best].enc;
+}
+
+/* When doing 'strict' detection, any string which is invalid in the candidate encoding
+ * is rejected. With non-strict detection, we just continue, but apply demerits for
+ * each invalid byte sequence */
+static const mbfl_encoding* mb_guess_encoding(unsigned char *in, size_t in_len, const mbfl_encoding **elist, unsigned int elist_size, bool strict)
+{
+	return mb_guess_encoding_for_strings(&in, &in_len, 1, elist, elist_size, strict);
 }
```

Comment:
If ``mb_guess_encoding_for_strings()`` can be changed to take const strings this funtion should also take const input.